csrc/transformer/inference/csrc/gelu.cu

```diff
@@ -432,71 +432,36 @@ template void launch_gptj_residual_add<__half>(__half*,
                                                int,
                                                int,
                                                cudaStream_t);
-
-__global__ void moe_res_matmul(float* residual,
-                               float* coef,
-                               float* mlp_out,
-                               int seq_len,
-                               int hidden_dim)
+template <typename T>
+__global__ void moe_res_matmul(T* residual, T* coef, T* mlp_out, int seq_len, int hidden_dim)
 {
-    unsigned tid = threadIdx.x;
-    float4* residual_cast = reinterpret_cast<float4*>(residual);
-    float4* coef_cast = reinterpret_cast<float4*>(coef);
-    float4* mlp_out_cast = reinterpret_cast<float4*>(mlp_out);
-
-    residual_cast += blockIdx.x * hidden_dim;
-    mlp_out_cast += blockIdx.x * hidden_dim;
-
-    float4* coef_cast2 = coef_cast + hidden_dim;
-
-    while (tid < hidden_dim) {
-        float4 res = residual_cast[tid];
-        float4 mlp = mlp_out_cast[tid];
-        float4 coef1 = coef_cast[tid];
-        float4 coef2 = coef_cast2[tid];
-        mlp.x = mlp.x * coef2.x + res.x * coef1.x;
-        mlp.y = mlp.y * coef2.y + res.y * coef1.y;
-        mlp.z = mlp.z * coef2.z + res.z * coef1.z;
-        mlp.w = mlp.w * coef2.w + res.w * coef1.w;
-        mlp_out_cast[tid] = mlp;
-        tid += blockDim.x;
-    }
-}
+    constexpr int granularity = 16;
+    constexpr int vals_per_access = granularity / sizeof(T);
 
-__global__ void moe_res_matmul(__half* residual,
-                               __half* coef,
-                               __half* mlp_out,
-                               int seq_len,
-                               int hidden_dim)
-{
-#ifdef HALF_PRECISION_AVAILABLE
-    unsigned tid = threadIdx.x;
-    float2* residual_cast = reinterpret_cast<float2*>(residual);
-    float2* mlp_out_cast = reinterpret_cast<float2*>(mlp_out);
-    float2* coef_cast = reinterpret_cast<float2*>(coef);
-    float2* coef_cast2 = coef_cast + hidden_dim;
-
-    residual_cast += blockIdx.x * hidden_dim;
-    mlp_out_cast += blockIdx.x * hidden_dim;
-
-    while (tid < hidden_dim) {
-        float2 res = residual_cast[tid];
-        float2 coef1 = coef_cast[tid];
-        float2 coef2 = coef_cast2[tid];
-        float2 data = mlp_out_cast[tid];
-        __half* data_h = reinterpret_cast<__half*>(&data);
-        __half* coef1_h = reinterpret_cast<__half*>(&coef1);
-        __half* coef2_h = reinterpret_cast<__half*>(&coef2);
-        __half* res_h = reinterpret_cast<__half*>(&res);
-        data_h[0] = res_h[0] * coef1_h[0] + data_h[0] * coef2_h[0];
-        data_h[1] = res_h[1] * coef1_h[1] + data_h[1] * coef2_h[1];
-        data_h[2] = res_h[2] * coef1_h[2] + data_h[2] * coef2_h[2];
-        data_h[3] = res_h[3] * coef1_h[3] + data_h[3] * coef2_h[3];
-
-        mlp_out_cast[tid] = data;
-        tid += blockDim.x;
+    unsigned tid = threadIdx.x * vals_per_access;
```

Comment:
This variable appears to be redundant since it is redefined on line 446.

Reply:
Done, thanks!